[2단계 - 자주 가는 음식점] 루나(오신정) 미션 제출합니다.

ShinjungOh opened the pull request:
# 🍱 자주 가는 음식점

안녕하세요 동키콩~~ 점심 뭐 먹지 미션 2단계를 완성해서 리뷰 요청 드립니다. 
이번에도 잘 부탁드려요! 🙇‍♀️

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [ ] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://shinjungoh.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

1. 타입스크립트 마이그레이션
이전에 사용해봐서 타입스크립트 적용은 수월했어요. **유효성 검증 로직**을 제외하고는 모두 타입스크립트로 마이그레이션 완료했어요. 유효성 검증을 타입스크립트로 변환하지 않은 이유는 `validateRestaurant.test.js` 단위 테스트 파일 때문이예요. jest에서 ts 파일을 검증하려면 별도의 설정이 추가적으로 필요하더라구요. 이 부분은 지금 당장 필요한 부분이 아니라고 생각해서 마이그레이션에서 제외했습니다.
타입가드를 이용해서 타입의 안전성을 확보하려고 했어요.
 any를 최대한 지양하려고 했는데 `localStorageUtils.ts`파일의 setLocalStorage 에서 1번 사용하고 있어요. 이 부분은 어떻게 타입을 지정해야 할지 모르겠어서 일단 any로 두었어요. 

2. 재활용 가능한 컴포넌트로 만들기
1단계 미션을 수행할 때, 2단계 미션을 고려하지 않고 설계했어요. 그래서 이번 2단계 미션을 구현할 때 도메인과 UI가 강하게 결합되어 있어서 너무 고생했습니다.😭 이전에 재활용이 어려웠던 모달 컴포넌트를 재활용이 가능하도록 UI와 도메인 기능을 분리해서 다음의 두가지 상황에 이용했습니다. `음식점 상세 모달`과 `음식점 추가 모달`을 Modal 컴포넌트를 이용해 재활용하는 방향으로 구현했습니다. 
마찬가지로 재활용 가능한 컴포넌트를 만들기 위해 Button 컴포넌트를 분리했었는데, 모달에서 버튼을 innerHTML로 만들고 있어서 Button 컴포넌트를 사용하지 않게 되었어요. 어디까지 분리를 하면 좋을지 고민하는 시간이 되었습니다. 

3. 함수형 프로그래밍 방식 사용
상태를 포함하며 관심사별로 묶여있는 **클래스**보다, **함수형**이 코드를 관리하기도 쉽고, 가독성이 좋다고 생각해서 함수형을 사용했습니다. 상태를 어디에 저장해야 할지 고민이 많이 되었는데, 상태 관리는 다음 3가지 방법을 이용해서 해결했어요.  
	- 1. 로컬 스토리지
	- 2. main.ts의 state
	- 3. RestaurantList의 localRestaurants

4. 로컬 스토리지와 restaurantManager
로컬스토리지를 쉽게 활용하기 위해 `localStorageUtils` 파일을 만들었어요. `restaurantManager` 파일에서 음식점 데이터를 localStorage를 통해 관리하고, 음식점에 고유 ID를 생성하기, 초기값 로드, 음식점 추가/삭제, 즐겨찾기 토글, 필터/정렬 기능을 담당하고 있어요.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

1. 새로운 음식점 추가 시 새로고침 하지 않으면 상세보기 모달을 볼 수 없는 문제
음식점을 추가하면 createRestaurantItem 함수로 `li` 요소를 만들고, 음식점 목록에 appendChild를 통해 추가되고 있어요. 
목록에서는 음식점 등록 시 바로 UI가 그려지고 있지만, 해당 요소를 클릭 시에 상세보기 모달이 뜨지는 않아요. 
시간이 부족해 이 부분을 해결하지 못하고 미션을 제출하지만, 원인을 파악해서 개선하려고 합니다.

2. `let localRestaurants`을 통한 상태 관리 방식 
RestaurantList 파일에서 `let localRestaurants: Restaurant[] = restaurants;` 코드로 상태를 관리하고 있습니다. 
로또 미션을 할 때 다음과 같은 프로그래밍 요구 사항이 있었어요.
![CleanShot 2025-03-17 at 15 27 55@2x](https://github.com/user-attachments/assets/8d759d6e-71bf-4956-94a8-df0df4fd8a5f)
상태 관리를 어떻게 해야할지 모르겠어서 임시방편으로 let을 이용하고 있는데, const만 사용해야 한다는 요구사항이 마음에 걸립니다. 이 부분을 개선해볼 수 있는 방법이 있을지 힌트를 받고 싶어요. 

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

ShinjungOh committed:
docs: README 기능 명세 추가

ShinjungOh committed:
feat: 음식점 목록에 즐겨찾기 아이콘 추가

ShinjungOh committed:
feat: 모든 음식점, 자주 가는 음식점 탭 UI 구현

ShinjungOh committed:
feat: 자주 가는 음식점 탭에서 셀렉트 영역 UI 숨기기

ShinjungOh committed:
refactor: Button 컴포넌트 분리

ShinjungOh committed:
feat: 음식점 상세 정보 조회 모달 UI 구현

ShinjungOh committed:
feat: JS 코드를 TS로 마이그레이션

ShinjungOh committed:
feat: 음식점 상세 모달 버튼 클릭시 동작 구현

ShinjungOh committed:
refactor: 모달 컴포넌트에서 Form 의존성 제거

ShinjungOh committed:
refactor: 변수명 오타 수정

ShinjungOh committed:
refactor: 모달 컴포넌트에서 상세 보기 모달 의존성 제거

- Modal 컴포넌트를 header, body, footer로 구조화

ShinjungOh committed:
refactor: Header 태그를 html 파일로 이동

ShinjungOh committed:
feat: 레스토랑 타입에 id, isFavorite 추가

- Distance, SortType 타입 정의

ShinjungOh committed:
feat: localStorage 유틸 함수 추가 및 음식점 추가, 삭제 구현

ShinjungOh committed:
feat: 자주 가는 음식점 등록, 제외 기능 구현

ShinjungOh committed:
feat: 자주 가는 음식점 탭 선택시 목록 조회 구현

ShinjungOh committed:
feat: 음식점, 탭, 카테고리 상태 추가

- RestaurantList 파일 분리
- 음식점 더미 데이터 추가
- 공통된 중복 로직 분리

ShinjungOh committed:
feat: 음식점 삭제 및 리렌더링 구현

ShinjungOh committed:
feat: 자주 가는 음식점 등록, 제외시 목록 리렌더링되도록 구현

ShinjungOh committed:
remove: 사용하지 않는 Button 컴포넌트 삭제

ShinjungOh committed:
docs: README 구현 완료 항목 체크리스트 업데이트

ShinjungOh committed:
refactor: 유효성 검사에서 카테고리, 거리 검증 로직 제거

ShinjungOh committed:
test: E2E 테스트로 사용자 시나리오 검증

ShinjungOh committed:
refactor: createModal 함수를 Modal.create 메소드로 변경

JUDONGHYEOK commented on src/restaurantManager.ts:
```diff
@@ -0,0 +1,104 @@
+
+  const localStorageRestaurantData = localStorageUtils.get("restaurants") || [];
+
+  if (localStorageRestaurantData.length === 0) {
+    const initialRestaurantsData = restaurantsData.map((restaurant) => {
+      const id = getUniqueRestaurantId();
+      return {
+        ...restaurant,
+        id,
+        isFavorite: false,
+      };
+    });
+    localStorageUtils.set("restaurants", initialRestaurantsData);
+  }
```
최초에 애플리케이션을 실행할 때 localStorage에만 데이터를 저장하고 렌더링할 때는 사용하지 않아서 새로고침해야 데이터가 나타나네요. 의도된 동작이신건지 궁금합니다!!

ShinjungOh replied:
의도하지 않았는데 새로고침 해야 데이터가 반영되도록 되어있네요...😂 
값이 바뀌면 바로 UI도 변경되도록 개선하겠습니다!

JUDONGHYEOK replied:
넵 현재는 정상동작 확인했습니다!

JUDONGHYEOK commented on src/main.ts:
```diff
@@ -0,0 +1,220 @@
+      return;
+    }
+
+    const restaurantItem = createRestaurantItem(newRestaurant);
+    restaurantList?.appendChild(restaurantItem);
```
생성된 restaurantItem에는 이벤트바인딩이 되지 않네요

ShinjungOh replied:
음식점 목록 호출, 상태를 업데이트하는 함수로 main의 코드를 개선했어요. 다시보니 전반적으로 main의 코드가 반복되고 구조화되지 않아서 공통으로 사용할 수 있는 함수를 만들어 관리하도록 했습니다.

JUDONGHYEOK replied:
main코드가 조금 비대해진 느낌이 나는데요. 개선이 필요할 것 같아요!!

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+    restaurantManager.toggleFavorite(id);
+    setRestaurant(updatedRestaurants);
+    localRestaurants = updatedRestaurants;
+    render(localRestaurants);
```
하나의 요소에 대한 favorite만으로 전체 요소를 렌더링하는 것이 로직을 간단하게 만들지만 성능상 비효율적일 수 있을것 같아요. 바뀐요소만 변경하는 것은 어떠세요?

ShinjungOh replied:
음.. 저는 편리하다고 생각했는데 프로젝트 규모가 커지고 코드가 복잡해지면 비효율적일 수 있겠네요. 
`data-id` 속성이 일치하는 요소를 찾아서 즐겨찾기 아이콘만 업데이트할 수 있도록 개선해볼게요!

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+  setRestaurant: (restaurants: Restaurant[]) => void,
+  el?: Element
+) => {
+  let localRestaurants: Restaurant[] = restaurants;
```
let으로 변수를 활용하는 것은 요구사항이 많아질수록 쉽게 오염될 수 있고 데이터 추적을 어렵게한다고 생각해요. 단순히 상태를 이용해보는 것도 좋을 것 같은데 어떻게 생각하세요?

ShinjungOh replied:
저도 고민했던 부분인데 부모 컴포넌트로에서 `restaurants` (음식점 배열)와 `setRestaurant` (상태 업데이트 함수)를 props로 전달받아 사용하도록 개선했어요!

JUDONGHYEOK replied:
상태를 관리할 경우에는 특히나 싱글톤인 경우에는 데이터의 오염방지를 위해서 접근제어가 중요하다고 생각해요 setRestaurant로 모든 레스토랑을 변경하는 함수를 넘겨주기 보다는 꼭 필요한 업데이트 로직만 넘겨주는 형태로 개선하면 어떨까요?

JUDONGHYEOK commented on src/main.ts:
```diff
@@ -0,0 +1,220 @@
+  category: "전체",
+  sortType: "name",
+  restaurants: [],
+};
```
디스크립션에서 말씀해주신 let관련 로직의 연장선상인데요. 싱글톤패턴의 모습을 적용하신 것 같은데요. 그렇다면 해당 상태를 사용해야하는 로직에서 추가로 let 변수를 활용하기보다는 싱글톤 변수를 직접적으로 사용하는 것도 괜찮을 것 같은데 어떻게 생각하시나요? 물론 필요한 값만 노출시키는 접근제어는 추가구현이 필요할 듯 하네요.

JUDONGHYEOK commented on src/utils/localStorageUtils.ts:
```diff
@@ -0,0 +1,45 @@
+const getLocalStorage = (key: string) => {
+  if (typeof window !== "undefined") {
+    const item = localStorage.getItem(key);
+    if (item) {
+      try {
+        return JSON.parse(item);
+      } catch (error) {
+        console.error(error);
+        return null;
+      }
+    }
+    return null;
+  }
+  return null;
+};
```
디스크립션에서 말씀해주신 부분인 것 같은데요. localStorage는 API와 유사하게 외부 의존성이라 타입을 코드에서 관리해줘야한다고 생각해요. 하나의 예로서 key-value 값을 아래와 같이 관리해줄 수 있을 것 같아요. set도 동일하게 구성해서 any를 제거할 수 있을 것 같네요
```typescript
interface ILocalStorageMap {
  'restaurants': Restaurant[];
}

const getLocalStorage = <T extends keyof ILocalStorageMap>(key: T) => {
  if (typeof window !== "undefined") {
    const item = localStorage.getItem(key);
    if (item) {
      try {
        return JSON.parse(item) as ILocalStorageMap[T];
      } catch (error) {
        console.error(error);
        return null;
      }
    }
    return null;
  }
  return null;
};
```

ShinjungOh replied:
알려주신 방법으로 사용해보았습니다! 제네릭은 아직 어려워서 잘 사용하지 않게 되는데, any를 제거하는 방법으로 유용할 것 같아요 👍

JUDONGHYEOK replied:
좋습니다 이제 첫 타입스크립트 미션이니 천천히 타입스크립트랑 친해져보아요~

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+  detailModal.showModal();
+
+  const modalFavoriteIcon = detailModal.querySelector(".favorite-icon");
+  if (modalFavoriteIcon instanceof HTMLImageElement) {
```
타입가드가 훌륭합니다!👍

JUDONGHYEOK commented on src/restaurantManager.ts:
```diff
@@ -0,0 +1,104 @@
+const addRestaurant = (restaurant: Restaurant) => {
+  const localStorageRestaurantData = localStorageUtils.get("restaurants") || [];
+
+  localStorageRestaurantData.push(restaurant);
+
+  localStorageUtils.set("restaurants", localStorageRestaurantData);
```
```suggestion
  localStorageUtils.set("restaurants", [...localStorageRestaurantData, restaurant]);
```

불변성을 지켜주는 것은 어떠세요??

ShinjungOh replied:
불변성을 유지시키면 부수효과를 줄일 수 있고 좋겠네요!

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+  };
+
+  const render = (restaurants: Restaurant[]) => {
+    if (el) {
+      el.innerHTML = "";
+    }
+
+    restaurants.forEach((restaurant: Restaurant) => {
+      const restaurantItem = createRestaurantItem(restaurant);
+
+      restaurantItem.addEventListener("click", (e) => {
+        if (
+          e.target instanceof HTMLImageElement &&
+          e.target.classList.contains("favorite-icon")
+        ) {
+          handleFavorite(restaurant.id);
+          return;
+        }
+        showRestaurantDetail(restaurant, handleDelete, handleFavorite);
+      });
+
+      el?.appendChild(restaurantItem);
+    });
+  };
```
성능적으로 개선해볼 수 있을 것 같은데요. 불필요한 DOM조작을 최소화하고 적절한 DOMAPI를 사용하는 것은 어떠세요?

ShinjungOh replied:
리뷰 덕분에 성능에 대해 고려해보는 시간이 되었어요. reflow와 repaint가 적게 일어날수록 좋다고 학습했는데 실제 코드를 작성할 때는 생각이 미치지 못한 것 같아요. 

아래 두가지 방법을 이용해 성능을 개선했어요.

1. fragment 사용
`createDocumentFragment`을 이용해 fragment를 만들었어요. 
매번 li 요소를 DOM에 직접 추가하는 대신에 fragment를 이용해 한 번만 부모 요소에 추가하도록 해서 성능을 향상시켰어요.

2. 이벤트 위임
`setAttribute`를 이용해 속성을 주고, 아이템마다 매번 등록되던 이벤트를 부모 요소에서 관리할 수 있도록 개선했어요.

JUDONGHYEOK commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,73 @@
+        options?.close.onClick();
+      }
+    });
+
```
모달이 열릴때 이벤트를 추가하고 닫힐때 이벤트를 제거해주고 있지는 않네요. 제거하는 로직이 추가되면 좋을 것 같은데 어떻게 생각하세요??

ShinjungOh replied:
재사용하는 모달이기 때문에 cleanup 함수를 추가해 이벤트 리스너를 제거했어요. 이벤트가 중복으로 등록될 가능성을 막아주었습니다.

JUDONGHYEOK replied:
관련 코멘트를 main코드에 적어놨으니 확인부탁드려요!

JUDONGHYEOK commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,73 @@
+          options
+            ? `<div class="modal-footer">
+          <div class="button-container">
+            <button type="button" id="modal-close-btn" class="button button--secondary text-caption">${options?.close.label}</button>
+            <button type="button" id="modal-submit-btn" class="button button--primary text-caption">${options?.submit.label}</button>
```
타입정의된 것으로 보면 옵셔널 체이닝이 필요없는 것 같아요

ShinjungOh replied:
해당 리뷰 반영했습니다👍

JUDONGHYEOK commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,73 @@
+            ? `<div class="modal-footer">
+          <div class="button-container">
+            <button type="button" id="modal-close-btn" class="button button--secondary text-caption">${options?.close.label}</button>
+            <button type="button" id="modal-submit-btn" class="button button--primary text-caption">${options?.submit.label}</button>
```
타입정의된 것으로 보면 옵셔널 체이닝이 필요없는 것 같아요

JUDONGHYEOK commented on src/components/Form.ts:
```diff
@@ -0,0 +1,51 @@
+      <div class="form-item form-item--required">
+        <label for="category" class="text-caption">카테고리</label>
+        <select name="category" id="category" required>
+          <option value="">선택해 주세요</option>
+          <option value="한식">한식</option>
+          <option value="중식">중식</option>
+          <option value="일식">일식</option>
+          <option value="양식">양식</option>
+          <option value="아시안">아시안</option>
+          <option value="기타">기타</option>
```
상수를 통해 반복되는 템플릿을 축소시킬 수 있을 것 같은데 어떻게 생각하세요??

ShinjungOh replied:
상수로 `CATEGORY_OPTIONS`과 `DISTANCE_OPTIONS`를 만들고, 이를 활용해서 반복되는 코드를 줄였어요!

JUDONGHYEOK replied:
좋습니다 또 Form은 너무 포괄적인 파일명이라고 생각해서 RestaurantsForm같은 파일명으로 바꿔줄 필요가 있을 것 같은데 어떻게 생각하세요?

ShinjungOh replied:
현재 컴포넌트들이 **Restaurant~** 라고 되어있어서 구체적인 관심사를 붙여 Form 컴포넌트의 이름도 `RestaurantForm`로 교체하고, 
기존에 `export { createForm }`라고 사용하는 방식에서 `export default` 로 내보내는 방식으로 개선했어요!

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+  render(localRestaurants);
+};
+
+const showRestaurantDetail = (
```
모듈의 크기가 커져 해당함수는 외부로 분리하시는 것은 어떠세요??

ShinjungOh replied:
`showRestaurantDetailModal` 함수를 분리해서 유지보수가 쉽고 관심사가 분리되도록 리팩토링 했어요.

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,143 @@
+    restaurants.forEach((restaurant: Restaurant) => {
+      const restaurantItem = createRestaurantItem(restaurant);
+
+      restaurantItem.addEventListener("click", (e) => {
```
renderRestaurantList에서 이벤트 바인딩을 해주는 것보다 restaurantItem에서 바인딩 하는 것이 더 책임이 명확하지 않을까요??

ShinjungOh replied:
https://github.com/woowacourse/javascript-lunch/pull/251#discussion_r2016054462

위에서 댓글로 이렇게 리팩토링 과정을 말씀드렸는데요,
RestaurantItem 컴포넌트에서 이벤트를 바인딩해주는 작업을 진행하고 나니 RestaurantList 컴포넌트의 코드도 깔끔해지고 관심사가 분리되어 훨씬 개선되었다는 생각이 드네요!

JUDONGHYEOK commented on cypress/e2e/spec.cy.js:
```diff
@@ -51,7 +51,120 @@ describe("Modal Test", () => {
+describe("View Restaurant List Test", () => {
+  beforeEach(() => {
+    cy.visit("http://localhost:5173", {
+      onBeforeLoad: (win) => {
+        win.localStorage.setItem(
+          "restaurants",
+          JSON.stringify(initialRestaurantsData)
+        );
+      },
```
테스트의 독립성을 구축하려 노력해주셨군요 훌륭하다고 생각합니다!!👍

ShinjungOh replied:
감사합니다😊

JUDONGHYEOK reviewed (changes requested):
안녕하세요 루나~~음식점 미션도 너무너무 수고많으셨어요!👍

타입스크립트 활용이 굉장히 인상적인 코드였어요. 제네릭을 활용하시는 모습이나 타입가드를 적절히 활용하시는 모습들도 보였고 타입을 좁히려고 노력하신 모습도 확인할 수 있었습니다. 인터페이스와 타입을 통해 코드 자체가 문서화 역할을 해주었던 것 같구요.
또 E2E테스트를 외부의존성이 추가되었음에도 성공적으로 구현해주셨네요!

다만 도메인 관리 측면에서는 아쉬운 점이 있었던 것 같은데요. 비즈니스 로직과 UI 로직의 경계가 다소 모호하며, 도메인 모델과 비즈니스 규칙이 UI에 혼재되어 있는 경우가 있었어요. 레스토랑과 관련된 비즈니스 규칙을 별도의 도메인 서비스나 값 객체로 분리하면 유지보수성이 더욱 향상될 거라고 생각해요!

질문주신 부분은 코멘트에 남겨두었으니 확인부탁드려요! 현재 기능이 정상적으로 동작하지 않는 부분이 있는 것 같아 RequestChanges하도록 하겠습니다!

JUDONGHYEOK commented:
@ShinjungOh 혹시 해당리뷰의 피드백은 언제 반영해주시는지 알 수 있을까요?

ShinjungOh commented:
@JUDONGHYEOK 
페어 미션이 겹쳐서 일정에 신경쓰지 못했어요. 죄송합니다🙇‍♀️
오늘 내일 작업해서 빠르게 리뷰 요청 보내겠습니다!

ShinjungOh committed:
refactor: 모달 cleanup 함수 추가

ShinjungOh committed:
refactor: 상수를 활용해 Form 컴포넌트의 select 중복 코드 개선

ShinjungOh committed:
refactor: showRestaurantDetailModal 함수 모듈화

ShinjungOh committed:
refactor: RestaurantList와 RestaurantItem 컴포넌트의 관심사 분리

ShinjungOh committed:
refactor: 음식점 목록 호출, 상태 업데이트하는 함수로 main의 코드를 개선

ShinjungOh committed:
refactor: let 제거 및 상태를 이용하도록 개선

ShinjungOh committed:
refactor: localStorageUtils의 타입 정의 및 any 제거

ShinjungOh committed:
fix: 새로고침 하지 않아도 음식점 등록, 제거되도록 수정

ShinjungOh committed:
fix: 음식점 목록에서 즐겨찾기 변경시 UI가 제대로 동작하지 않는 이슈 해결

- UI 변경시 updateRestaurantList를 통해 변경하도록 개선
- handleFavorite 함수로 즐겨찾기 여부 변경되도록 구현

ShinjungOh commented:
# 🍱 자주 가는 음식점 - 리팩토링

안녕하세요 동키콩~!! 다시 돌아오기까지 오랜 시간이 걸렸네요😭😂
미리 늦어진다고 일정을 공유드렸어야 하는데 그러지 못한 점 사과드려요. 

지난 미션이니까 금방 해치우자는 생각보다는, 내가 무엇 때문에 유독 런치 미션이 힘들게 느껴졌나 돌아보면서 
특히 스스로 저에게 부족했다고 느껴지는 부분을 중점으로 개선해 보려고 노력했어요.
저는 `구조와 설계`하는 것에 약한데, 영화 미션을 하면서도 계속 처음부터 어긋난 방향으로 시작을 해서 구조를 갈아엎는데에 시간을 많이 쏟게 됐어요. 리팩토링 하면서 스스로 배우는 것도 있고 리뷰로 가이드를 주셔서 도움이 많이 되었습니다🙇‍♀️

특히 동키콩도 피드백 주셨던 **도메인 관리 측면에서 비즈니스 로직과 UI 로직의 경계가 다소 모호하며, 도메인 모델과 비즈니스 규칙이 UI에 혼재되어 있는 경우**가 이번 리팩토링을 하며 풀었던 과제였어요. 제가 제 코드를 리팩토링 하다보니, 유지보수가 굉장히 어려운 코드라는걸 많이 느꼈거든요.

남겨주신 피드백 하나씩 살펴보고 고민하는 시간을 가졌어요. 리팩토링에서 변경된 사항과 고민했던 부분을 남겨봅니다. 
점진적으로 리팩토링 해나가는 과정에서 피드백 주셨던 부분에 대해 반영한 코드가 바뀌기도 했는데, 제가 코멘트로 남겨두었던 댓글과 코드가 달라졌을 수 있어 말씀드립니다!

<br>

## ⚙️ 주요 변경 사항

### 1. let을 제거하고 state를 이용하기

[refactor: let 제거 및 상태를 이용하도록 개선](https://github.com/woowacourse/javascript-lunch/pull/251/commits/a7a4a900a10cebd6931abf9c4ce3c41130abbfa3)
[refactor: 음식점 목록 호출, 상태 업데이트하는 함수로 main의 코드를 개선](https://github.com/woowacourse/javascript-lunch/pull/251/commits/1a899164a30b0805ad2962dc031aff1095e524cc)

프로그래밍 요구사항이었던 'let 사용하지 않기'를 해결했어요. 상태를 통해서 데이터를 뿌려주도록 했고, main에서는 상태를 업데이트할 수 있도록 개선했어요.


### 2 . 컴포넌트별 관심사 분리

[refactor: showRestaurantDetailModal 함수 모듈화](https://github.com/woowacourse/javascript-lunch/pull/251/commits/481da75f1ab6cc918a9a5358776728be593d4171)
[refactor: RestaurantList와 RestaurantItem 컴포넌트의 관심사 분리](https://github.com/woowacourse/javascript-lunch/pull/251/commits/16d13ce74abb56058f394ebc3e2b51b95f234d21)

너무 길어지는 컴포넌트를 관심사별로 분리하고 함수를 모듈화 했어요.
`Modal`과, Modal 컴포넌트를 사용하는 `RestaurantDetailModal` 컴포넌트로 모달을 관리하고 있어요.


### 3. 새로고침해야 데이터가 리렌더링되는 문제 해결 

[fix: 새로고침 하지 않아도 음식점 등록, 제거되도록 수정](https://github.com/woowacourse/javascript-lunch/pull/251/commits/017785ff1aed4b358b714ef853cca3d593b5146d)
[fix: 음식점 목록에서 즐겨찾기 변경시 UI가 제대로 동작하지 않는 이슈 해결](https://github.com/woowacourse/javascript-lunch/pull/251/commits/b1ff20ab4b74b8e8e1b8afdd306fae951a189959)

가장 문제가 되었던 부분인데요, 새로고침해야 변경된 데이터가 UI에 반영되는 문제를 개선했어요. 
자주 가는 음식점(즐겨찾기) 버튼을 눌렀을 때 값이 바로 변경되지 않는 버그도 있어서 해결했습니다!

### 4. 기타 변경사항

- 타입스크립트 any 제거 및 타입 관련 피드백 적용 
- 모달에 이벤트 cleanup 함수 추가
- 상수를 활용해 Form 컴포넌트의 select에서 사용되는 중복 코드 개선

<br>

## 테스트 통과 스크린샷

![CleanShot 2025-03-28 at 18 51 54@2x](https://github.com/user-attachments/assets/2ed27717-06ae-4499-97c6-1cc67aa44119)

JUDONGHYEOK commented on src/main.ts:
main파일이 너무 비대해진 것 같다는 생각도 드는데요. 200줄이 넘는 코드가 하나의 파일로 구성되어 있어서 로직을 쉽게 파악하기 어렵게 만드는 것 같아요. 따라서 책임에 따른 분리가 필요한 것 같은데요. 제가 생각하기엔 이벤트바인딩을 컴포넌트에서 담당하고 상태를 변경하는 도메인 로직들을 하나의 객체로도 묶을 수 있을 것 같아요

JUDONGHYEOK commented on src/main.ts:
```diff
@@ -0,0 +1,216 @@
+  body?.append(addRestaurantModal);
+
+  const addRestaurantModalButton = header?.querySelector(".gnb__button");
+  addRestaurantModalButton?.addEventListener("click", () => {
+    addRestaurantModal.showModal();
+  });
```
기존에 정의된 addRestaurantModal을 show, close하고 있기 때문에 버튼에 할당된 클린업함수들이 실행되고 버튼들에 새로운 이벤트들이 할당되지 않아요 따라서 등록하기, 취소하기 이벤트는 애플리케이션 생명주기에서 한번밖에 할당이 되지 않아 두 개이상의 레스토랑등록이나 취소가 불가능한 상황입니다.

[beforeToggle](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/beforetoggle_event)이나 toggle같은 이벤트들도 있는데요 현재는 크로미움 기반의 브라우저에만 정상동작해서 호환성에 맞게 적절하게 사용하면 좋겠네요

JUDONGHYEOK commented on src/validateRestaurant.js:
```diff
@@ -13,20 +13,13 @@ const validateRestaurant = (newRestaurant, restaurantNames) => {
-  }
-
-  if (newRestaurant.description.length > VALIDATE_SETTINGS.descriptionMaxLength) {
+  if (
+    newRestaurant.description.length > VALIDATE_SETTINGS.descriptionMaxLength
+  ) {
```
validate가 사라진 이유가 있을까요? 카테고리나 거리가 필수값인 것처럼 보이는데 현재는 설정하지 않고도 등록이 가능합니다

ShinjungOh replied:
제가 개발하다가 지워야겠다고 생각한 이유가 있었는데 지금 보니 왜 그런지 기억이 나지 않네요💀
안그래도 왜 오류창이 뜨지 않는지 의문이었는데 유효성 검사 코드 다시 추가했습니다😂

JUDONGHYEOK commented on src/components/RestaurantList.ts:
```diff
@@ -0,0 +1,93 @@
+  filter,
+}: RestaurantListProps) => {
+  const render = () => {
+    el.innerHTML = "";
```
replaceChildren을 사용하면 비워주지 않아도 될것 같아요!

ShinjungOh replied:
`el.innerHTML = "";` 코드를 삭제하고, appendChild를 `replaceChildren` 메소드로 교체해서 
최종적으로 `el.replaceChildren(fragment);` 코드로 개선했어요!

JUDONGHYEOK reviewed (changes requested):
안녕하세요 루나! 리뷰반영을 훌륭하게 해주셔서 감사해요. 확실히 애플리케이션의 구조가 안정화된 느낌이 납니다. 
어프로브 하고 싶지만 기능 요구사항에서 문제가 되는 부분이 있어서요. 해당 부분만 수정해주시면 어프로브하러 다시 오도록 하겠습니다.

추가 코멘트는 참고용으로만 봐주셔도 되고 모든 기능이 우선적으로 동작하도록 먼저 수정해주시면 감사하겠습니다!

ShinjungOh committed:
refactor: Form 컴포넌트를 RestaurantForm으로 네이밍 변경, export 방식 변경

ShinjungOh committed:
refactor: RestaurantList컴포넌트의 appendChild를 replaceChildren 메소드로 교체

ShinjungOh committed:
refactor: dom 함수 생성 및 적용

- document.querySelector 함수 생성 및 적용
- createElement 함수 생성 및 적용

ShinjungOh committed:
fix: 음식점 추가 아이콘이 보이지 않는 이슈 해결

ShinjungOh committed:
feat: 카테고리, 거리 유효성 검사 추가

ShinjungOh committed:
refactor: RestaurantAddModal 컴포넌트 분리

- main의 음식점 추가 모달 관련 코드를 RestaurantAddModal로 이동
- dom.ts 파일의 $ 함수에 제네릭 타입 추가
- 거리 유효성 검사 시 0의 타입을 숫자로 변경
- Modal의 cleanUp 함수 제거

ShinjungOh committed:
refactor: 음식점 추가 모달 제거 로직 추가

ShinjungOh committed:
refactor: main의 코드를 관심사에 따라 restaurantElement, restaurantStore로 분리

ShinjungOh commented:
# 🍱 자주 가는 음식점 - 리팩토링 ver.2

안녕하세요 동키콩!

이번이 진짜 ‘최종_마지막_리팩토링.ts’이 되길 바라며😂 리뷰를 최대한 반영해서 개선했어요! 

요구사항이 지켜지지 않던 기능도 제대로 돌아가도록 고치고, main의 코드도 줄이는 것을 이번 리팩토링의 중점으로 두었어요. 아직 개선해야할 부분이 보이지만 일단 기능 위주로 리팩토링 했습니다. 

<br>

## ⚙️ 주요 변경 사항

### 1. 음식점 추가 모달이 제대로 작동하지 않는 이슈 해결

이전 리팩토링 시에는 상세 모달은 제대로 작동하는데, 음식점 추가 모달이 제대로 작동하지 않았어요. 두 모달의 차이를 보니 상세 모달은 매번 모달을 지우고 새로 생성하는데, 추가 모달은 하나의 모달을 계속 사용하고 있어서 문제가 발생했더라구요. 음식점 추가 모달도 매번 새로 생성해주는 방식으로 교체했습니다.

### 2. main 파일 개선

너무 길어진 main 파일에서 `RestaurantAddModal` 부분을 분리해 새로운 컴포넌트를 생성했어요. 
`restaurantElement`와 `restaurantStore` 두개의 파일을 생성해 main에 혼재되어 있던 코드를 관심사별로 묶어주었어요. 

### 3. dom.ts 유틸함수 생성 및 적용

dom.ts 유틸함수를 만들어서 `document.querySelector`와 `createElement` 만들어줬어요. 반복되는 코드를 줄이고, HTML요소에는 $표기를 해서 element임을 확실히 표기해 주었어요. `createElement`에서는 알려주신 제네릭 타입을 사용했어요!

JUDONGHYEOK reviewed (approved):
안녕하세요 루나!! 모든 요구사항을 만족하는 코드인 것을 확인하여 어프로브 하도록 하겠습니다!!

긴 기간동안 너무 고생많으셨습니다~ 앞으로의 다른 미션들도 화이팅입니다!!
